### PR TITLE
Data model prose

### PR DESCRIPTION
**Pull request by sfsinger19103**
Trying to make the document more accessible to CDOS and other readers

### CODE REVIEW

File: specs/pvs/corla.pvs [hunk 1]
```diff
@@ -1,33 +1,78 @@
...
+%%% case of Colorado, it facilitates running risk limiting audits
+%%% across all counties in the state simultaneously.
+%%%
+%%% The Colorado election law pertaining to election audits is
+%%% \href{http://law.justia.com/codes/colorado/2016/title-1/general-primary-recall-and-congressional-vacancy-elections/article-7/part-5/section-1-7-515/}
+%%% {Section 1-7-515, C.R.S.}  and the rules pertaining to such are
+%%% found in the "Revised Draft of Proposed Rules" whose latest
+%%% version is labeled CCR 1505-1 and dated July 6, 2017. Section 25,
+%%% Post-Election Audit, of that document is the salient portion of
+%%% that document. We will call this document RDPR-6-Jul-2017 for the
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: that document _for our purposes_

File: specs/pvs/corla.pvs [hunk 2]
```diff
@@ -40,16 +85,14 @@
 %%% informal specifications which will be extracted using a shell script
 %%% into a well-typed FAFESSL informal domain model specification.
 %%%
-%%% As mentioned above, we write formal specifications using a
-%%% literate style, a la Knuth. Doing so permits us to document our
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: "literate" and "literate programming" are technical terms, but don't really sound like it. Thus I think it makes sense to retain the reference to Knuth, or perhaps, better, something like https://en.wikipedia.org/wiki/Literate_programming, to help clarify for readers what is going on.

Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: I agree. The reference could be moved - perhaps to the first place where "literate" is mentioned - but should stay somewhere.

File: specs/pvs/corla.pvs [hunk 3]
```diff
@@ -1,33 +1,78 @@
...
-%%% of the system delivered to the Colorado Department of State. The prose
-%%% annotations should help Colorado Department of State elections domain experts
+%%% of the elements of the specification. 
+%%% The specification documents our
+%%% thinking and the system design, from domain modeling and
+%%% engineering all of the way to formal specification and
+%%% verification.
+%%%
+%%% We write the formal specification using a
+%%% literate style, allowing the production of typeset books
```
Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: This would be a good place for the reference to Knuth and literate programming.

File: specs/pvs/corla.pvs [hunk 4]
```diff
@@ -1,33 +1,78 @@
...
+%%% and interactive hypertext directly from the specification.
+%%% The prose
+%%% annotations will help Colorado Department of State elections domain experts
 %%% and technical staff to assess the completeness and correctness of the
-%%% ColoradoRLA system/tool.
+%%% ColoradoRLA tool.
```
Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: Is it the RLA Tool or the ColoradoRLA Tool?

File: specs/pvs/corla.pvs [hunk 5]
```diff
@@ -1,33 +1,78 @@
...
+%%% in Java and runs on servers hosted by CDOS. The client is
+%%% implemented in TypeScript and JavaScript and runs in modern web
+%%% browsers.
+%%% CDOS provides
+%%% the hosting servers, network, and a number of services
+%%% in support of the tools deployment and management.
 %%%
 %%% This document is written in a literate style using the PVS
 %%% theorem prover. Writing system specifications in this fashion is
 %%% called (formal) domain engineering. Our formal domain model is
```
Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: If we're removing "our" in the line below, we should remove it here too.

File: specs/pvs/corla.pvs [hunk 6]
```diff
@@ -203,9 +209,13 @@ END information_systems
 %%% requirements. Preferably redundant servers would be hosted in
 %%% separate facilities, with separate and independent power systems
 %%% and networks.
-%%%
```
Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: Are you intending to merge two paragraphs here? (that's what's happening with the LaTeX document when you delete this line)

File: specs/pvs/corla.pvs [hunk 7]
```diff
@@ -203,9 +209,13 @@ END information_systems
...
 %%% An Election Night Reporting (ENR) server is mentioned repeatedly
 %%% in the RDPR-6-Jul-2017, so we model it here.
+%%% The next few lines of the specification encode the process of establishing
```
Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: Similarly, is this (line 214) now meant to be a paragraph on its own?

File: specs/pvs/corla.pvs [hunk 8]
```diff
@@ -269,14 +276,16 @@ BEGIN
...
   candidate: TYPE FROM person
 
-%%% Elections focus on contests, each of which represents a set of
-%%% choices that voters can make. Making a legal choice ranges
+%%% Elections focus on contests, each of which represents a 
+%%% choice that voters can make. Making a legal choice ranges
 %%% enormously across the Earth, from marking a vote for a candidate
 %%% by filling in a single bubble to enumerating a total order on all
-%%% choices in a contest. Across a given ballot, each contest has a
+%%% choices in a contest. Within any single elction, each contest has a
```
Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: election

File: specs/pvs/corla.pvs [hunk 9]
```diff
@@ -1,33 +1,78 @@
...
+%%% the hosting servers, network, and a number of services
+%%% in support of the tools deployment and management.
 %%%
 %%% This document is written in a literate style using the PVS
 %%% theorem prover. Writing system specifications in this fashion is
 %%% called (formal) domain engineering. Our formal domain model is
-%%% written in PVS's higher-order logic (HOL). Our informal domain
-%%% model is written in the Free \& Fair System Specification Language,
+%%% written in PVS's higher-order logic (HOL). There is a corresponding informal domain
+%%% model written in the Free \& Fair System Specification Language,
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: Isn't that Free & Fair*_Election_ System Specification Language, here and a few other places? (quoting https://github.com/FreeAndFair/BON)?  Or are you re-thinking the name and positioning it as more general? Though the "E" in FAFSSL is handy for pronunciation. It could be "Free & Fair Extended System Specification Language I guess.

Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: Well, it really _is_ more general, given that the specification language itself has nothing to do with election systems (heck, depending on where the funding for actually building the tooling for it comes from, it might be GSSL - the Galois System Specification Language - which is how we've referred to it the past few years). I'd actually drop the A from the acronym and make it "FFSSL", if I had to choose something this minute.

Comment (dmzimmerman) [inline specs/pvs/corla.pvs]: (or "KZSSL" :P )

Comment (nealmcb) [inline specs/pvs/corla.pvs]: Beware the "Freakishly Fast and Simple Scripting Language" project.
KZSSL is more fun to pronounce, and doesn't seem to conflict with anything ;)

File: specs/pvs/corla.pvs [hunk 10]
```diff
@@ -69,12 +112,16 @@
...
 
-%%% Our formal domain model is also annotated with structured comments
-%%% in precise natural language using a standard set of annotations.
-%%% A shell script processes these annotations and generates a
-%%% well-typed informal FAFESSL specification.
+%+BEGIN background
+
+%%% \section{Formal Model Description}
+
+%%%\subsection{Basic Technical Infrastructure}
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: Missing space after %%%, here and one other place.

File: specs/pvs/corla.pvs [hunk 11]
```diff
@@ -92,35 +139,9 @@
...
-
-% @todo kiniry Add a discussion of our glossary and its relationship
-% to this specification.
-
-% @todo kiniry Define C.R.S. and explain the other sources used to
-% derive all concepts in this specification.
-
 % @todo Explain how refinement works and how the reader can understand
 % exactly what is built vs. what is just designed.
+% @review kiniry what's the point of including these annotations? I don't see them used in the document @sfsinger
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: Which annotations do you mean? The @todo above, or the Literate PVS notes below? Both seem used and useful to me.

File: specs/pvs/corla.pvs [hunk 12]
```diff
@@ -136,23 +157,8 @@ END fafessl
...
-%%% in Java and runs on servers hosted by CDOS. The client is
-%%% implemented in TypeScript and JavaScript and runs in modern web
-%%% browsers.
-%%%
-%%% As we are hosting the RLA Tool in CDOS's system, they are
-%%% providing the hosting servers, network, and a number of services
-%%% in support of the tools deployment and management.
-%%%
-%%% A database is used to store all of the data relevant to
+%%%% A database is used to store all of the data relevant to
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: Extra %

File: specs/pvs/corla.pvs [hunk 13]
```diff
@@ -269,14 +276,16 @@ BEGIN
...
 %%% enormously across the Earth, from marking a vote for a candidate
 %%% by filling in a single bubble to enumerating a total order on all
-%%% choices in a contest. Across a given ballot, each contest has a
+%%% choices in a contest. Within any single elction, each contest has a
 %%% unique name.
 % @todo kiniry Need to break the cycle between the contest type and
-% the ballots theory.
+% the ballots theory. sfsinger suggests: contests are grouped into elecitons,
+% not ballots. So define contests first. Then a ballot is created to 
+% represent a set of contests.
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: The work of the NIST/EAC Interoperability Working group, Election Modeling subgroup, might be useful. (Yeah, pretty late to bring this up, but I just noticed their recent progress, FWIW).  The seem to have a BPMN2 model, pictures, XPDL, xml, who-knows-what at https://github.com/usnistgov/ElectionModeling.  See also
http://collaborate.nist.gov/voting/bin/view/Voting/ElectionModeling
Also the UML for the https://github.com/usnistgov/CastVoteRecords work.

File: specs/pvs/corla.pvs [hunk 14]
```diff
@@ -394,24 +409,37 @@ END election
...
+% as subsets of "contests". A ballot question is a contest between "yes" and "no".
 %%%
-%%% Ballots are stored by one of several possible means. Some ballot
+%%% There are a variety of ways to store ballots. Some ballot
 %%% storage containers are secure, others are not. Some facilitate an
 %%% easy means by which to find a particular ballots, others do not.
 
 % @relation box inherit storage_container
 % @relation bin inherit storage_container
 % @relation ballot_box inherit storage_container
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: Is there supposed to be a connection between these notions of boxes vs bins and the notions of efficiency or secure vs insecure storage? Why are we distinguishing these types of storage?

File: specs/pvs/corla.pvs [hunk 15]
```diff
@@ -625,11 +667,11 @@ END ballots_collections
 
 %+BEGIN ifr
 
-%%% \section{Instructions, Forms, and Reports}
+%%% \subsection{Instructions, Forms, and Reports}
 
 %%% Elections have all kinds of different published instructions,
 %%% forms, and reports. We enumerate a few here that are specific to
-%%% Colorado, and the RLA Tool in particular.
+%%% Colorado, and relevant the RLA Tool in particular.
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: relevant _to_

File: specs/pvs/corla.pvs [hunk 16]
```diff
@@ -758,20 +802,23 @@ BEGIN
...
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
-%%% \section{Cryptography}
+%%% \subsection{Cryptography}
 
-%%% We will add additional cryptographic primitives here as required.
+% We will add additional cryptographic primitives here as required.
+% @review kiniry dmz Is it correct to say that the cryptography is used
+% (via hashing) to verify that the ballot manifests and cvrs stored in the 
+% database are actually the ones uploaded by the Counties?
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: The hash confirms first of all that the file upload proceeds without any corruption, when it is checked after receipt.  It can also be used by counties, CDOS and the public to confirm that everyone is talking about exactly the same file, after it is downloaded again. It looks like they will be stored as rather large binary objects in Postgres. I hope it and our deployment / design can handle objects that are pushing a gigabyte in size, though we won't see that in this state, this year.

File: specs/pvs/corla.pvs [hunk 17]
```diff
@@ -781,17 +828,18 @@ END cryptography
 
 %%% \subsection{Randomness}
 
-%%% High quality randomness is key to the legitimacy of any
-%%% risk-limiting audit.
+%%% Choosing a random sample of ballots is key to the legitimacy of any
+%%% risk-limiting audit. Legitimacy depends on a high-quality process for
+%%% the random samples. The Colorado RLA Tool will produce a random order 
+%%% of all the ballots once and for all; then random samples can be created
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: => of all the ballots _for each county_ once and for all; then random samples _of the necessary size_ can be created

File: specs/pvs/corla.pvs [hunk 18]
```diff
@@ -1001,23 +1050,29 @@ END rlas
...
 %%% syntactically written as comma-separated values, and other times
 %%% in plain English. Some CVRs use a election-specific encoding
 %%% scheme to represent choices (e.g., a '1' means "John Doe"); others
 %%% use plain English.
 %%%
+%%% CVRs may (and in the case of CVRs exported from Colorado's Dominion system, do)
+%%% contain information about the ballot beyond the voter choices. For 
+%%% example, ballot style, precinct id and ballot type (in-person vs. mail)
+%%% may be included.
+%%%
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: `CountingGroup` is the one that are "Mail" or "In Person". `BallotType` is the ballot style, which includes a precinct number which is a big problem from a CORA standpoint. We would prefer that all counties upload a CVR which has all the precinct information left out: `PrecinctPortion`, the precinct part of `BallotType`, and `ImprintedId`.

File: specs/pvs/corla.pvs [hunk 19]
```diff
@@ -1001,23 +1050,29 @@ END rlas
...
 %%% CDOS requirements mandate that CVRs are exported and uploaded to
 %%% the RLA back-end as CSV files by county officials using their
 %%% voting systems and the RLA Tool, respectively.
 
 % @todo kiniry For a given election, there must be a 1-1
 % correspondence between its ballots and its CVRs.
+% @review kiniry between its ballots and the data-rows of its CVR(s), right?
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: I don't think we can actually always expect either a 1-1 correspondence between paper ballots and CVRs (given phantom ballots) or manifest entries and either of those (human error, to be handled also via phantom ballots).

File: specs/pvs/corla.pvs [hunk 20]
```diff
@@ -1043,14 +1098,18 @@ END cast_vote_records
...
 %%% The user interfaces of the system are the visible interactive
 %%% parts of the application. There are three different user
 %%% interfaces in the RLA Tool. The precise names of these interfaces
 %%% are still under discussion; one is for CDOS personnel responsible
 %%% for the audit at the state-level, one is for county personnel at
 %%% the county-level, and one is for audit board members.
+%%% Note that all these user interfaces both receive and 
+%%% provide information.
+%%% Because the public dashboard only pushes information out, 
+%%% we do not consider it a "user interface".
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: I think we ought to add the public dashboard to the list.  It public dashboard will not change the database, but will accept some sorts of interaction and/or input, whether mouse clicks or searches or the like.

File: specs/pvs/corla.pvs [hunk 21]
```diff
@@ -1283,8 +1348,6 @@ BEGIN
 
 % @trace RDPR-6-Jul-2017 Section 25.2.2(K)
 
-%%% \subsection{Random Selection of Ballots for Audit}
```
Comment (nealmcb) [inline specs/pvs/corla.pvs]: Did you mean to move this up above this line?
`%%% The state-wide dashboard will allow the Secretary of State`

File: specs/pvs/corla.pvs (hunk 9 above)
Comment (sfsinger19103) [inline specs/pvs/corla.pvs]: I'll leave action on these comments to others.

File: specs/pvs/corla.pvs (hunk 17 above)
Comment (sfsinger19103) [inline specs/pvs/corla.pvs]: One random *order* of all ballots state-wide allows random samples restricted to counties: for a random sample of size n for County X, simply take the first n ballots from County X that appear in the statewide random order.

Comment (nealmcb) [inline specs/pvs/corla.pvs]: That would work, but I'm thinking it would make it harder for the public to verify the selection, since now they have to include all manifests when they do so, in some canonical order among counties, and then do their own winnowing of the overall list to generate what is selected for a given county. They couldn't e.g. just feed a single county manifest (with a bit of tweaking) into Philip's tool and confirm the sample that way. I'll have to think about this....

File: specs/pvs/corla.pvs (hunk 7 above)
Comment (nealmcb) [inline specs/pvs/corla.pvs]: It does seem better with a paragraph break here, and for the comment above.

**Comment by kiniry** [comment]
I will review and work on this as soon as PR #108 is merged.

**Comment by kiniry** [comment]
PR #108 is now merged, so I'll look at this now.

File: specs/pvs/corla.pvs (hunk 14 above)
Comment (kiniry) [inline specs/pvs/corla.pvs]: I was distinguishing them because (a) I had witnessed different storage mechanisms in the field, (b) I understand there is not a uniform mechanism in CO, and (c) some mechanisms are secure and provide a chain-of-custody evidence and others are not.  This latter point matters wrt the meaning of `secure` and `verified`.

**Comment by nealmcb** [comment]
Some comments from my continuing review of what was merged:

change 'undervote' to 'blank vote' most everwhere

at some point fix comment around audited_ballot_by_contest_list_table
 https://github.com/FreeAndFair/ColoradoRLA/pull/108/commits/24ab0327e1b82a2737963a7adc98496b70ad3193#r130158627

> +%%% This domain model is based upon these statues, its associated

=> statutes

> +%%% traceability relation, then the reason for its introduce should be

=> introduction

> %%% apply a hash value to the CVR export file using the hash value

should we confirm this hash? need to know what it is - sha256?

File: specs/pvs/corla.pvs (hunk 13 above)
Comment (kiniry) [inline specs/pvs/corla.pvs]: Making sure that our model is consistent with federal standards someday is a good thing, but we don't have time or resources for it now. What consistent usually means is that there is a relation we can define between types in the models that is information preserving from the simpler model to the more complex one.

File: specs/pvs/corla.pvs (hunk 16 above)
Comment (kiniry) [inline specs/pvs/corla.pvs]: I'm in alignment with everything that @nealmcb stated here.  Thinking carefully about foolproof means by which 100s of megabyte uploads can happen reliably and securely is important.  Historically I see government agencies purchasing commercial products for such services but they don't work as well as the off-the-shelf open source stuff we use.  E.g., I would think that a secure rsync, or at least a restartable sftp, is wiser than anything else in this regards.

File: specs/pvs/corla.pvs (hunk 19 above)
Comment (kiniry) [inline specs/pvs/corla.pvs]: Understood. In a perfect world, they would be 1-1, but that'll basically never happen.  And it seems like different vendors do different things about the relationship between CVRs and multi-page ballots.

File: specs/pvs/corla.pvs (hunk 20 above)
Comment (kiniry) [inline specs/pvs/corla.pvs]: I agree with you @nealmcb.  A UI is a UI, regardless of if it is read-only or not. We need to specify events and scenarios even for the public dashboard, in this regards.

File: specs/pvs/corla.pvs (hunk 16 above)
Comment (nealmcb) [inline specs/pvs/corla.pvs]: A first easy step is to support automatic compression. They give us a csv, we compress it (as a .zip?), upload it, process it (ideally on-the-fly) and store it. Then offer the compressed version for download.

Comment (nealmcb) [inline specs/pvs/corla.pvs]: I think CDOS does have some sort of secure ftp infrastructure for counties, which we might be able to leverage, by letting them point to a reference to the uploaded file which we could pull from.
